Add new web3modal, Reuse active websocket  connection for sign v2

0xAsimetriq opened the pull request:
## Description

Replaces legacy `qrcode-modal` with standalone version of new web3modal. Added guard to not create / initiate new WalletConnect provider every time `connect` method is called in v2, as same socket can be reused.

## Additional Information

- [x] I read the [contributing docs](/wagmi-dev/references/blob/main/.github/CONTRIBUTING.md) (if this is your first contribution)

## Tests
- Tested v1 implementation with metamask / rainbow wallet using docs example page
- Tested v2 implementation with new web3modal and universal-provider via Trust wallet using docs example
- Tested closing the modal and throwing an error
- Tested rejecting connection within wallet and throwing an error

Your ENS/address: asimetriq.eth

jxom commented on packages/connectors/src/walletConnect.ts:
```diff
@@ -80,13 +101,13 @@ export class WalletConnectConnector extends Connector<
       const provider = await this.getProvider({
         chainId: targetChainId,
-        create: true,
+        create: isV2 && this.#provider ? false : true,
```
Does targeting a `chainId` still work? Can we fix that if it does not?

0xAsimetriq replied:
`chainId` still worked, but good catch here as this still created new modal instance. I had to add secondary check within getProvider for `!this.#provider || create` and use `as UniversalProvider` as you can see in latest commit. Not the prettiest solution probably, but had to satisfy TypeScript. I am up to do a bigger refactor for `getProvider` to split checks for chainId and creating wc instances if thats ok? Could make things a bit nicer.

jxom replied:
Yep, if we can refactor `getProvider` that would be a good win!

0xAsimetriq commented:
@jxom I addressed your comments, plus ran the tests, everything seems to be ok, however I'll keep this in draft for now until I can figure out how to test everything properly in wagmi docs site (questions in our chat). Gancho also added some new changes to universal-provider that clean up and optimise websocket subscriptions, will be adding that new version in latter on as well during docs site tests.

0xAsimetriq commented:
@jxom This one should be ready for review now. Updated universal-provider and managed to set up example docs to test everything. New modal is now in place for v2 (still using old one for v1 as it is pre bundled into ethereum-provider).

I attempted to refactor that `getProvider` method, but didn't end up with anything cleaner to be fair, however made it less nested in terms of if statements. Will return to this once we deprecate v1, as a lot can be removed then.

jxom commented on packages/connectors/src/walletConnect.ts:
```diff
@@ -266,41 +281,44 @@ export class WalletConnectConnector extends Connector<
+      ).default
+      this.#provider = new WalletConnectProvider({
+        ...this.options,
+        qrcode: false,
```
```suggestion
```

We want to leave this as-is for v1 right?

0xAsimetriq replied:
Yes, sorry, was left over while I was experimenting with adding new modal to v1 as well, but it will require far to many changes as old one is coupled within `ethereum-provider`, so we will be tackling that within it and once ready I will open another PR to update `ethereum-provider` 👍

jxom reviewed (commented):
Looking pretty good now! Just one minor comment.

0xAsimetriq commented:
@jake as per our chat on improving loading times, I added these changes

Main issue was actually with that initial loading of `universal-provider`, so what I did is start loading / initialising it in constructor. Similarly, we start loading / initialising web3modal there as well. This gives chance to start importing associated files before user clicks on a "connect button".

I added few utils that check for readiness of everything just in case user manages to click on "connect button" before everything is loaded up (i.e. 3g networks).

In terms of preloading data within modal, it is already relatively optimised i.e we use cloudflare workers to deliver it at the edge / cache is in place / we only fetch first few wallets that are needed for initial display. While we are also preloading images, these are non-blocking in terms of opening modal. Loading ui here is usually on a "connect button" itself.

jxom commented on packages/connectors/src/walletConnect.ts:
```diff
@@ -77,13 +85,48 @@ export class WalletConnectConnector extends Connector<
+  async #waitForReady(condition: () => boolean) {
+    if (condition()) return true
+    await new Promise((resolve) => {
+      const interval = setInterval(() => {
+        if (condition()) {
+          clearInterval(interval)
+          resolve(true)
+        }
+      }, 100)
```
Can we instead rely on the pattern we have in `getProvider` where we check if it has already been instantiated, otherwise create it instead of relying on an interval? Would it be possible to remove `createUniversalProvider` altogether and instead call `getProvider` (for v2) in the constructor? This will also remove the need for `isUniversalProviderReady` as well.

If we need to check instantiation for web3modal, we should do the same pattern as `getProvider`. This would remove the need for `isWeb3ModalReady`.

0xAsimetriq replied:
Other reason why I did it this way was to avoid double initialisations i.e. if for whatever reason `getProvider` called from constructor is still "loading" and due to that we call it again in say `connect`. With web3modal v2 this would append second `<w3m-modal />` web-component into dom.
